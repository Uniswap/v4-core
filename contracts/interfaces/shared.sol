// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.10;

/// @notice Represents a change in the pool's balance of token0 and token1.
struct BalanceDelta {
    int256 amount0;
    int256 amount1;
}
