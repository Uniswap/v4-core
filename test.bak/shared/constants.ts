export const MIN_TICK = -887272
export const MAX_TICK = 887272
export const MAX_TICK_SPACING = 32767
export const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000'
