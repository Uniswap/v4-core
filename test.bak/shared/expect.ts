import { expect, use } from 'chai'
import { solidity } from 'ethereum-waffle'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'

use(solidity)
use(jestSnapshotPlugin())

export { expect }
