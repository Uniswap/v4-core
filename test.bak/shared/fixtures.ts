import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { TestERC20 } from '../../typechain/TestERC20'

interface TokensFixture {
  currency0: TestERC20
  currency1: TestERC20
  token2: TestERC20
}

export async function tokensFixture(): Promise<TokensFixture> {
  const tokenFactory = await ethers.getContractFactory('TestERC20')
  const tokenA = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenB = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenC = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20

  const [currency0, currency1, token2] = [tokenA, tokenB, tokenC].sort((tokenA, tokenB) =>
    tokenA.address.toLowerCase() < tokenB.address.toLowerCase() ? -1 : 1
  )

  return { currency0, currency1, token2 }
}
