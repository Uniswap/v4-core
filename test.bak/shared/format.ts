import { Decimal } from 'decimal.js'
import { BigNumberish } from 'ethers'

export function formatTokenAmount(num: BigNumberish): string {
  return new Decimal(num.toString()).dividedBy(new Decimal(10).pow(18)).toPrecision(5)
}

export function formatPrice(price: BigNumberish): string {
  return new Decimal(price.toString()).dividedBy(new Decimal(2).pow(96)).pow(2).toPrecision(5)
}
