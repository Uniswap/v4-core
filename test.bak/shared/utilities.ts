import bn from 'bignumber.js'
import { BigNumber, BigNumberish, Contract, ContractTransaction, utils, Wallet } from 'ethers'
import { ethers } from 'hardhat'

export const MaxUint128 = BigNumber.from(2).pow(128).sub(1)

export const getMinTick = (tickSpacing: number) => Math.ceil(-887272 / tickSpacing) * tickSpacing
export const getMaxTick = (tickSpacing: number) => Math.floor(887272 / tickSpacing) * tickSpacing

export const MIN_SQRT_RATIO = BigNumber.from('4295128739')
export const MAX_SQRT_RATIO = BigNumber.from('1461446703485210103287273052203988822378723970342')

export enum FeeAmount {
  LOW = 500,
  MEDIUM = 3000,
  HIGH = 10000,
}

export const TICK_SPACINGS: { [amount in FeeAmount]: number } = {
  [FeeAmount.LOW]: 10,
  [FeeAmount.MEDIUM]: 60,
  [FeeAmount.HIGH]: 200,
}

export function expandTo18Decimals(n: number): BigNumber {
  return BigNumber.from(n).mul(BigNumber.from(10).pow(18))
}

bn.config({ EXPONENTIAL_AT: 999999, DECIMAL_PLACES: 40 })

// returns the sqrt price as a 64x96
export function encodeSqrtPriceX96(reserve1: BigNumberish, reserve0: BigNumberish): BigNumber {
  return BigNumber.from(
    new bn(reserve1.toString())
      .div(reserve0.toString())
      .sqrt()
      .multipliedBy(new bn(2).pow(96))
      .integerValue(3)
      .toString()
  )
}

export function getPositionKey(address: string, lowerTick: number, upperTick: number): string {
  return utils.keccak256(utils.solidityPack(['address', 'int24', 'int24'], [address, lowerTick, upperTick]))
}

export function getPoolId({
  currency0,
  currency1,
  fee,
  tickSpacing,
  hooks,
}: {
  currency0: string | Contract
  currency1: string | Contract
  fee: number
  tickSpacing: number
  hooks: string | Contract
}): string {
  return utils.keccak256(
    utils.defaultAbiCoder.encode(
      ['address', 'address', 'uint24', 'int24', 'address'],
      [
        typeof currency0 === 'string' ? currency0 : currency0.address,
        typeof currency1 === 'string' ? currency1 : currency1.address,
        fee,
        tickSpacing,
        typeof hooks === 'string' ? hooks : hooks.address,
      ]
    )
  )
}

export type SwapFunction = (
  amount: BigNumberish,
  to: Wallet | string,
  sqrtPriceLimitX96?: BigNumberish
) => Promise<ContractTransaction>
export type SwapToPriceFunction = (sqrtPriceX96: BigNumberish, to: Wallet | string) => Promise<ContractTransaction>
export type ModifyPositionFunction = (
  tickLower: BigNumberish,
  tickUpper: BigNumberish,
  liquidityDelta: BigNumberish
) => Promise<ContractTransaction>
export type DonateFunction = (amount0: BigNumberish, amount1: BigNumberish) => Promise<ContractTransaction>

interface HookMask {
  beforeInitialize: boolean
  afterInitialize: boolean
  beforeModifyPosition: boolean
  afterModifyPosition: boolean
  beforeSwap: boolean
  afterSwap: boolean
  beforeDonate: boolean
  afterDonate: boolean
}

/**
 * Creates a 20 byte mask for the given hook configuration
 */
export function createHookMask({
  beforeInitialize,
  afterInitialize,
  beforeModifyPosition,
  afterModifyPosition,
  beforeSwap,
  afterSwap,
  beforeDonate,
  afterDonate,
}: HookMask): string {
  let result: BigNumber = BigNumber.from(0)
  if (beforeInitialize) result = result.add(BigNumber.from(1).shl(159))
  if (afterInitialize) result = result.add(BigNumber.from(1).shl(158))
  if (beforeModifyPosition) result = result.add(BigNumber.from(1).shl(157))
  if (afterModifyPosition) result = result.add(BigNumber.from(1).shl(156))
  if (beforeSwap) result = result.add(BigNumber.from(1).shl(155))
  if (afterSwap) result = result.add(BigNumber.from(1).shl(154))
  if (beforeDonate) result = result.add(BigNumber.from(1).shl(153))
  if (afterDonate) result = result.add(BigNumber.from(1).shl(152))
  return utils.hexZeroPad(result.toHexString(), 20)
}

/**
 * Returns a wallet whose first transaction will create a contract satisfying the leading
 * bytes required by hookMask. If provided, the mnemonic argument short-circuits our search
 * to save time.
 */
export function getWalletForDeployingHookMask(hookMask: HookMask, mnemonic?: string): [Wallet, string] {
  const startingString = createHookMask(hookMask).slice(0, 4)
  let wallet: Wallet = mnemonic ? ethers.Wallet.fromMnemonic(mnemonic) : ethers.Wallet.createRandom()
  let contractAddress: string | undefined

  while (contractAddress === undefined) {
    const prospectiveContractAddress = utils.getContractAddress({ from: wallet.address, nonce: 0 })
    if (prospectiveContractAddress.slice(0, 4).toLowerCase() === startingString) {
      contractAddress = prospectiveContractAddress
    } else {
      // if, for whatever reason, we generate a bad address but a mnemonic was provided,
      // it's stale and we surface the error
      if (mnemonic) throw Error('Stale mnemonic')
      wallet = ethers.Wallet.createRandom()
    }
  }

  return [wallet, contractAddress]
}
