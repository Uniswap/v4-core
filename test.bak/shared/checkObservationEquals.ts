import { BigNumber, BigNumberish } from 'ethers'
import { expect } from './expect'

// helper function because we cannot do a simple deep equals with the
// observation result object returned from ethers because it extends array
export default function checkObservationEquals(
  {
    tickCumulative,
    blockTimestamp,
    initialized,
    secondsPerLiquidityCumulativeX128,
  }: {
    tickCumulative: BigNumber
    secondsPerLiquidityCumulativeX128: BigNumber
    initialized: boolean
    blockTimestamp: number
  },
  expected: {
    tickCumulative: BigNumberish
    secondsPerLiquidityCumulativeX128: BigNumberish
    initialized: boolean
    blockTimestamp: number
  }
) {
  expect(
    {
      initialized,
      blockTimestamp,
      tickCumulative: tickCumulative.toString(),
      secondsPerLiquidityCumulativeX128: secondsPerLiquidityCumulativeX128.toString(),
    },
    `observation is equivalent`
  ).to.deep.eq({
    ...expected,
    tickCumulative: expected.tickCumulative.toString(),
    secondsPerLiquidityCumulativeX128: expected.secondsPerLiquidityCumulativeX128.toString(),
  })
}
