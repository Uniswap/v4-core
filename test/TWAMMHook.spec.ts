import { createFixtureLoader } from 'ethereum-waffle'
import { BigNumber, Wallet } from 'ethers'
import hre, { ethers } from 'hardhat'
import snapshotGasCost from '@uniswap/snapshot-gas-cost'
import { TWAMMHook, PoolManager, PoolModifyPositionTest, PoolSwapTest, TestERC20 } from '../typechain'
import { MAX_TICK_SPACING } from './shared/constants'
import { expect } from './shared/expect'
import { tokensFixture } from './shared/fixtures'
import {
  createHookMask,
  encodeSqrtPriceX96,
  expandTo18Decimals,
  getMaxTick,
  getMinTick,
  getPoolId,
} from './shared/utilities'
import { inOneBlock } from './shared/inOneBlock'

function nIntervalsFrom(timestamp: number, interval: number, n: number): number {
  return timestamp + (interval - (timestamp % interval)) + interval * (n - 1)
}

type OrderKey = {
  owner: string
  expiration: number
  zeroForOne: boolean
}

// TODO: temporary hack for precision errors affecting token balances
const EXTRA_TOKENS = 20

describe('TWAMM Hook', () => {
  let wallets: Wallet[]
  let wallet: Wallet
  let twamm: TWAMMHook
  let poolManager: PoolManager
  let modifyPositionTest: PoolModifyPositionTest
  let token0: TestERC20
  let token1: TestERC20

  /**
   * We are using hardhat_setCode to deploy the twamm, so we need to replace all the immutable references
   * @param poolManagerAddress the address of the pool manager
   * @param expirationInterval the aexpiration interval
   */
  async function getDeployedTWAMMCode(poolManagerAddress: string, expirationInterval: number): Promise<string> {
    const artifact = await hre.artifacts.readArtifact('TWAMMHook')
    const fullyQualifiedName = `${artifact.sourceName}:${artifact.contractName}`
    const debugArtifact = await hre.artifacts.getBuildInfo(fullyQualifiedName)
    const immutableReferences =
      debugArtifact?.output?.contracts?.[artifact.sourceName]?.[artifact.contractName]?.evm?.deployedBytecode
        ?.immutableReferences
    if (!immutableReferences) throw new Error('Could not find immutable references')
    if (Object.keys(immutableReferences).length !== 2) throw new Error('Unexpected immutable references length')
    const addressRefs = immutableReferences[Object.keys(immutableReferences)[1]]
    const expirationRefs = immutableReferences[Object.keys(immutableReferences)[0]]
    let bytecode: string = artifact.deployedBytecode
    const paddedTo32Address = '0'.repeat(24) + poolManagerAddress.slice(2)
    const expirationHex = expirationInterval.toString(16)
    const paddedTo32Expiration = '0'.repeat(64 - expirationHex.length) + expirationHex
    for (const { start, length } of addressRefs) {
      if (length !== 32) throw new Error('Unexpected immutable reference length')
      bytecode = bytecode.slice(0, start * 2 + 2) + paddedTo32Address + bytecode.slice(2 + start * 2 + length * 2)
    }
    for (const { start, length } of expirationRefs) {
      if (length !== 32) throw new Error('Unexpected immutable reference length')
      bytecode = bytecode.slice(0, start * 2 + 2) + paddedTo32Expiration + bytecode.slice(2 + start * 2 + length * 2)
    }
    return bytecode
  }

  const fixture = async ([wallet]: Wallet[]) => {
    const twammFactory = await ethers.getContractFactory('TWAMMHook')

    const poolManagerFactory = await ethers.getContractFactory('PoolManager')
    const swapTestFactory = await ethers.getContractFactory('PoolSwapTest')
    const modifyPositionTestFactory = await ethers.getContractFactory('PoolModifyPositionTest')
    const tokens = await tokensFixture()
    const manager = (await poolManagerFactory.deploy()) as PoolManager

    const twammHookAddress = createHookMask({
      beforeInitialize: true,
      afterInitialize: false,
      beforeModifyPosition: true,
      afterModifyPosition: false,
      beforeSwap: true,
      afterSwap: false,
      beforeDonate: false,
      afterDonate: false,
    })

    await hre.network.provider.send('hardhat_setCode', [
      twammHookAddress,
      await getDeployedTWAMMCode(manager.address, 10_000),
    ])

    const twamm: TWAMMHook = twammFactory.attach(twammHookAddress) as TWAMMHook

    const swapTest = (await swapTestFactory.deploy(manager.address)) as PoolSwapTest
    const modifyPositionTest = (await modifyPositionTestFactory.deploy(manager.address)) as PoolModifyPositionTest

    for (const token of [tokens.token0, tokens.token1]) {
      // TODO: HACK!! send some extra tokens so that the current precision errors do not revert tests
      token.transfer(twamm.address, EXTRA_TOKENS)
      for (const spender of [swapTest, modifyPositionTest, twamm]) {
        await token.connect(wallet).approve(spender.address, ethers.constants.MaxUint256)
      }
    }

    return {
      manager,
      swapTest,
      modifyPositionTest,
      tokens,
      twamm,
    }
  }

  let loadFixture: ReturnType<typeof createFixtureLoader>
  before('create fixture loader', async () => {
    wallets = await (ethers as any).getSigners()
    wallet = wallets[0]

    loadFixture = createFixtureLoader(wallets)
  })

  let poolKey: {
    token0: string
    token1: string
    fee: number
    tickSpacing: number
    hooks: string
  }

  beforeEach('deploy twamm', async () => {
    ;({
      twamm,
      manager: poolManager,
      tokens: { token0, token1 },
      modifyPositionTest,
    } = await loadFixture(fixture))
    poolKey = {
      token0: token0.address,
      token1: token1.address,
      fee: 0,
      hooks: twamm.address,
      tickSpacing: MAX_TICK_SPACING,
    }
  })

  let snapshotId: string

  beforeEach('check the pool is not initialized', async () => {
    const { sqrtPriceX96 } = await poolManager.getSlot0(poolKey)
    expect(sqrtPriceX96, 'pool is not initialized').to.eq(0)
    // it seems like the waffle fixture is not working correctly (perhaps due to hardhat_setCode), and if we don't
    // do this and revert in afterEach, the pool is already initialized
    snapshotId = await hre.network.provider.send('evm_snapshot')
  })

  afterEach('revert', async () => {
    // see the beforeEach hook
    await hre.network.provider.send('evm_revert', [snapshotId])
  })

  describe('#beforeInitialize', async () => {
    it('initializes the twamm', async () => {
      expect((await twamm.twammStates(getPoolId(poolKey))).lastVirtualOrderTimestamp).to.equal(0)
      await poolManager.initialize(poolKey, encodeSqrtPriceX96(1, 1))
      expect((await twamm.twammStates(getPoolId(poolKey))).lastVirtualOrderTimestamp).to.equal(
        (await ethers.provider.getBlock('latest')).timestamp
      )
    })
  })

  describe('#executeTWAMMOrders', async () => {
    let poolKey: any
    let orderKey0: any
    let orderKey1: any
    let latestTimestamp: number

    beforeEach(async () => {
      latestTimestamp = (await ethers.provider.getBlock('latest')).timestamp

      poolKey = {
        token0: token0.address,
        token1: token1.address,
        fee: 0,
        hooks: twamm.address,
        tickSpacing: 10,
      }
      orderKey0 = {
        zeroForOne: true,
        owner: wallet.address,
        expiration: nIntervalsFrom(latestTimestamp, 10_000, 3),
      }
      orderKey1 = {
        zeroForOne: false,
        owner: wallet.address,
        expiration: nIntervalsFrom(latestTimestamp, 10_000, 3),
      }

      await ethers.provider.send('evm_setNextBlockTimestamp', [nIntervalsFrom(latestTimestamp, 10_000, 1)])
      await poolManager.initialize(poolKey, encodeSqrtPriceX96(1, 1))
      latestTimestamp = (await ethers.provider.getBlock('latest')).timestamp

      await modifyPositionTest.modifyPosition(poolKey, {
        tickLower: getMinTick(10),
        tickUpper: getMaxTick(10),
        liquidityDelta: expandTo18Decimals(1),
      })
    })

    it('gas with no initialized ticks', async () => {
      await inOneBlock(latestTimestamp + 100, async () => {
        await twamm.submitLongTermOrder(poolKey, {
          amountIn: expandTo18Decimals(1),
          ...orderKey0,
        })
        await twamm.submitLongTermOrder(poolKey, {
          amountIn: expandTo18Decimals(10),
          ...orderKey1,
        })
      })
      await ethers.provider.send('evm_setNextBlockTimestamp', [latestTimestamp + 10_000])
      await snapshotGasCost(twamm.executeTWAMMOrders(poolKey))
    })

    it('gas crossing 1 initialized tick', async () => {
      await modifyPositionTest.modifyPosition(poolKey, {
        tickLower: -200,
        tickUpper: 200,
        liquidityDelta: expandTo18Decimals(1),
      })
      await inOneBlock(latestTimestamp + 100, async () => {
        await twamm.submitLongTermOrder(poolKey, {
          amountIn: expandTo18Decimals(1),
          ...orderKey0,
        })
        await twamm.submitLongTermOrder(poolKey, {
          amountIn: expandTo18Decimals(10),
          ...orderKey1,
        })
      })
      await ethers.provider.send('evm_setNextBlockTimestamp', [latestTimestamp + 10_000])
      await snapshotGasCost(twamm.executeTWAMMOrders(poolKey))
    })

    it('gas crossing 2 initialized ticks', async () => {
      await modifyPositionTest.modifyPosition(poolKey, {
        tickLower: -200,
        tickUpper: 200,
        liquidityDelta: expandTo18Decimals(1),
      })
      await modifyPositionTest.modifyPosition(poolKey, {
        tickLower: -2000,
        tickUpper: 2000,
        liquidityDelta: expandTo18Decimals(1),
      })

      await inOneBlock(latestTimestamp + 100, async () => {
        await twamm.submitLongTermOrder(poolKey, {
          amountIn: expandTo18Decimals(1),
          ...orderKey0,
        })
        await twamm.submitLongTermOrder(poolKey, {
          amountIn: expandTo18Decimals(10),
          ...orderKey1,
        })
      })

      await ethers.provider.send('evm_setNextBlockTimestamp', [latestTimestamp + 10_000])
      await snapshotGasCost(twamm.executeTWAMMOrders(poolKey))
    })
  })

  describe('end-to-end integration', () => {
    let start: number
    let expiration: number
    let key: any
    let orderKey0: OrderKey
    let orderKey1: OrderKey
    let liquidityBalance: BigNumber

    describe('when TWAMM crosses a tick', () => {
      beforeEach('set up pool', async () => {
        const latestTimestamp = (await ethers.provider.getBlock('latest')).timestamp
        const amountLiquidity = expandTo18Decimals(1)

        start = nIntervalsFrom(latestTimestamp, 10_000, 1)
        expiration = nIntervalsFrom(latestTimestamp, 10_000, 3)

        orderKey0 = {
          zeroForOne: true,
          owner: wallet.address,
          expiration,
        }
        orderKey1 = {
          zeroForOne: false,
          owner: wallet.address,
          expiration,
        }

        key = {
          token0: token0.address,
          token1: token1.address,
          fee: 0,
          hooks: twamm.address,
          tickSpacing: 10,
        }
        await poolManager.initialize(key, encodeSqrtPriceX96(1, 1))

        // 1) Add liquidity balances to AMM
        await modifyPositionTest.modifyPosition(key, {
          tickLower: getMinTick(10),
          tickUpper: getMaxTick(10),
          liquidityDelta: amountLiquidity,
        })
        await modifyPositionTest.modifyPosition(key, {
          tickLower: -20000,
          tickUpper: 20000,
          liquidityDelta: amountLiquidity,
        })

        liquidityBalance = await token0.balanceOf(poolManager.address)
      })

      it('balances clear properly w/ token1 excess', async () => {
        const amountLTO0 = expandTo18Decimals(1)
        const amountLTO1 = expandTo18Decimals(10)

        await ethers.provider.send('evm_setAutomine', [false])

        // 2) Add order balances to TWAMM
        await twamm.submitLongTermOrder(key, {
          amountIn: amountLTO0,
          ...orderKey0,
        })
        await twamm.submitLongTermOrder(key, {
          amountIn: amountLTO1,
          ...orderKey1,
        })

        await ethers.provider.send('evm_mine', [start])
        await ethers.provider.send('evm_setAutomine', [true])
        await ethers.provider.send('evm_setNextBlockTimestamp', [expiration + 1000])

        const prevBalance0 = await token0.balanceOf(twamm.address)
        const prevBalance1 = await token1.balanceOf(twamm.address)

        await twamm.executeTWAMMOrders(key)

        const newBalance0 = await token0.balanceOf(twamm.address)
        const newBalance1 = await token1.balanceOf(twamm.address)

        const earningsToken1 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey0)
        const earningsToken0 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey1)

        // TODO: precision error of 3-6 wei :(
        // With both trades expiring, the final balance of TWAMM should be equal to the earnings able to be claimed
        // for each TWAMM orderPools
        expect(newBalance0.sub(EXTRA_TOKENS)).to.eq(earningsToken0.sub(4))
        expect(newBalance1.sub(EXTRA_TOKENS)).to.eq(earningsToken1.sub(3))
      })

      it('balances clear properly w/ token0 excess', async () => {
        const amountLTO0 = expandTo18Decimals(10)
        const amountLTO1 = expandTo18Decimals(1)

        await ethers.provider.send('evm_setAutomine', [false])

        // 2) Add order balances to TWAMM
        await twamm.submitLongTermOrder(key, {
          amountIn: amountLTO0,
          ...orderKey0,
        })
        await twamm.submitLongTermOrder(key, {
          amountIn: amountLTO1,
          ...orderKey1,
        })

        await ethers.provider.send('evm_mine', [start])
        await ethers.provider.send('evm_setAutomine', [true])
        await ethers.provider.send('evm_setNextBlockTimestamp', [expiration + 1000])

        const prevBalance0 = await token0.balanceOf(twamm.address)
        const prevBalance1 = await token1.balanceOf(twamm.address)

        await twamm.executeTWAMMOrders(key)

        const newBalance0 = await token0.balanceOf(twamm.address)
        const newBalance1 = await token1.balanceOf(twamm.address)

        const earningsToken1 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey0)
        const earningsToken0 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey1)

        // TODO: precision error of 3-6 wei :(
        // With both trades expiring, the final balance of TWAMM should be equal to the earnings able to be claimed
        // for each TWAMM order
        expect(newBalance0.sub(EXTRA_TOKENS)).to.eq(earningsToken0.sub(5))
        expect(newBalance1.sub(EXTRA_TOKENS)).to.eq(earningsToken1.sub(6))
      })
    })

    describe('when TWAMM crosses no ticks', () => {
      it('clears all balances appropriately when trading against a 0 fee AMM', async () => {
        const latestTimestamp = (await ethers.provider.getBlock('latest')).timestamp
        const amountLiquidity = expandTo18Decimals(1)
        const amountLTO0 = expandTo18Decimals(1)
        const amountLTO1 = expandTo18Decimals(10)

        const start = nIntervalsFrom(latestTimestamp, 10_000, 1)
        const expiration = nIntervalsFrom(latestTimestamp, 10_000, 3)

        const orderKey0 = {
          zeroForOne: true,
          owner: wallet.address,
          expiration,
        }
        const orderKey1 = {
          zeroForOne: false,
          owner: wallet.address,
          expiration,
        }

        const key = {
          token0: token0.address,
          token1: token1.address,
          fee: 0,
          hooks: twamm.address,
          tickSpacing: 10,
        }
        await poolManager.initialize(key, encodeSqrtPriceX96(1, 1))

        // 1) Add liquidity balances to AMM
        await modifyPositionTest.modifyPosition(key, {
          tickLower: getMinTick(10),
          tickUpper: getMaxTick(10),
          liquidityDelta: amountLiquidity,
        })

        await ethers.provider.send('evm_setAutomine', [false])

        // 2) Add order balances to TWAMM
        await twamm.submitLongTermOrder(key, {
          amountIn: amountLTO0,
          ...orderKey0,
        })
        await twamm.submitLongTermOrder(key, {
          amountIn: amountLTO1,
          ...orderKey1,
        })

        await ethers.provider.send('evm_mine', [start])
        await ethers.provider.send('evm_setAutomine', [true])
        await ethers.provider.send('evm_setNextBlockTimestamp', [expiration + 1000])

        const prevBalance0 = await token0.balanceOf(twamm.address)
        const prevBalance1 = await token1.balanceOf(twamm.address)

        await twamm.executeTWAMMOrders(key)

        const newBalance0 = await token0.balanceOf(twamm.address)
        const newBalance1 = await token1.balanceOf(twamm.address)

        const earningsToken1 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey0)
        const earningsToken0 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey1)

        // TODO: precision error of 3-6 wei :(
        // With both trades expiring, the final balance of TWAMM should be equal to the earnings able to be claimed
        // for each TWAMM order
        expect(newBalance0.sub(EXTRA_TOKENS)).to.eq(earningsToken0.sub(3))
        expect(newBalance1.sub(EXTRA_TOKENS)).to.eq(earningsToken1.sub(4))
      })
    })

    describe('single pool sell', () => {
      beforeEach('set up pool', async () => {
        const latestTimestamp = (await ethers.provider.getBlock('latest')).timestamp
        const amountLiquidity = expandTo18Decimals(1)

        start = nIntervalsFrom(latestTimestamp, 10_000, 1)
        expiration = nIntervalsFrom(latestTimestamp, 10_000, 3)

        orderKey0 = {
          zeroForOne: true,
          owner: wallet.address,
          expiration,
        }

        key = {
          token0: token0.address,
          token1: token1.address,
          fee: 0,
          hooks: twamm.address,
          tickSpacing: 10,
        }
        await poolManager.initialize(key, encodeSqrtPriceX96(1, 1))

        // 1) Add liquidity balances to AMM
        await modifyPositionTest.modifyPosition(key, {
          tickLower: getMinTick(10),
          tickUpper: getMaxTick(10),
          liquidityDelta: amountLiquidity,
        })
        await modifyPositionTest.modifyPosition(key, {
          tickLower: -2000,
          tickUpper: 2000,
          liquidityDelta: amountLiquidity,
        })
        await modifyPositionTest.modifyPosition(key, {
          tickLower: -3000,
          tickUpper: 3000,
          liquidityDelta: amountLiquidity,
        })
      })

      describe('when crossing two ticks', () => {
        it('crosses both ticks properly', async () => {
          await ethers.provider.send('evm_setNextBlockTimestamp', [start])

          await twamm.submitLongTermOrder(key, {
            amountIn: expandTo18Decimals(5),
            ...orderKey0,
          })

          await ethers.provider.send('evm_setNextBlockTimestamp', [expiration + 300])
          await twamm.executeTWAMMOrders(key)

          const newBalance0 = await token0.balanceOf(twamm.address)
          const newBalance1 = await token1.balanceOf(twamm.address)

          const earningsToken1 = await twamm.callStatic.claimEarningsOnLongTermOrder(key, orderKey0)

          // TODO: precision error of 11 wei :(
          // The only order is completed so the new balance should equal the earnings to be collected
          expect(newBalance1.sub(EXTRA_TOKENS)).to.equal(earningsToken1.sub(11))
        })
      })
    })
  })
})
