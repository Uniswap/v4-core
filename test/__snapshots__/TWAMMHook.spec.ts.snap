// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TWAMM Hook #claimTokens gas 1`] = `
Object {
  "calldataByteLength": 100,
  "gasUsed": 35815,
}
`;

exports[`TWAMM Hook #executeTWAMMOrders when both order pools are selling gas crossing 1 initialized tick 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 528550,
}
`;

exports[`TWAMM Hook #executeTWAMMOrders when both order pools are selling gas crossing 2 initialized ticks 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 638590,
}
`;

exports[`TWAMM Hook #executeTWAMMOrders when both order pools are selling gas with no initialized ticks 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 430301,
}
`;

exports[`TWAMM Hook #executeTWAMMOrders when only one pool is selling gas crossing 1 initialized tick 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 341377,
}
`;

exports[`TWAMM Hook #executeTWAMMOrders when only one pool is selling gas crossing 2 initialized ticks 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 363643,
}
`;

exports[`TWAMM Hook #executeTWAMMOrders when only one pool is selling gas crossing no initialized tick 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 316818,
}
`;

exports[`TWAMM Hook #submitLongTermOrder gas 1`] = `
Object {
  "calldataByteLength": 292,
  "gasUsed": 139941,
}
`;

exports[`TWAMM Hook #updateLongTermOrder gas 1`] = `
Object {
  "calldataByteLength": 292,
  "gasUsed": 230045,
}
`;

exports[`TWAMM Hook bytecode size 1`] = `26051`;
